### ISSUE TITLE
COPY no longer honoring .. (folder above) in paths

### ISSUE THREAD

**Issue by thoraxe** [containers/libpod#5035]
**Is this a BUG REPORT or FEATURE REQUEST? (leave only one on its own line)**

/kind bug

**Description**
Building a Dockerfile where the COPY statement that references the path "../file" no longer works

**Steps to reproduce the issue:**

1. Create a Dockerfile with something like:
```
COPY ../foo ./
```

2. `podman build -t blahblah .`

**Describe the results you received:**
```
$ pwd
/home/thoraxe/Documents/xxx/container-allinone
$ podman build -t blahblah .
...
...
STEP 4: COPY ../Gemfile ../Gemfile.lock ./
Error: error dry-running "COPY ../Gemfile ../Gemfile.lock ./": no files found matching "/home/thoraxe/Documents/xxx/container-allinone/Gemfile": no such file or directory
```

**Describe the results you expected:**
Build successfully

**Additional information you deem important (e.g. issue happens only occasionally):**

**Output of `podman version`:**

```
Version:            1.7.0
RemoteAPI Version:  1
Go Version:         go1.13.5
OS/Arch:            linux/amd64
```

**Output of `podman info --debug`:**

```
debug:
  compiler: gc
  git commit: ""
  go version: go1.13.5
  podman version: 1.7.0
host:
  BuildahVersion: 1.12.0
  CgroupVersion: v2
  Conmon:
    package: conmon-2.0.9-2.fc31.x86_64
    path: /usr/bin/conmon
    version: 'conmon version 2.0.9, commit: 7d46f3e7711aa3578488284ae2f98b447658f086'
  Distribution:
    distribution: fedora
    version: "31"
  IDMappings:
    gidmap:
    - container_id: 0
      host_id: 1000
      size: 1
    - container_id: 1
      host_id: 100000
      size: 65536
    uidmap:
    - container_id: 0
      host_id: 1000
      size: 1
    - container_id: 1
      host_id: 100000
      size: 65536
  MemFree: 3229913088
  MemTotal: 32841293824
  OCIRuntime:
    name: crun
    package: crun-0.10.6-1.fc31.x86_64
    path: /usr/bin/crun
    version: |-
      crun version 0.10.6
      spec: 1.0.0
      +SYSTEMD +SELINUX +APPARMOR +CAP +SECCOMP +EBPF +YAJL
  SwapFree: 16104812544
  SwapTotal: 16471027712
  arch: amd64
  cpus: 8
  eventlogger: journald
  hostname: t490s-festive-local
  kernel: 5.4.13-201.fc31.x86_64
  os: linux
  rootless: true
  slirp4netns:
    Executable: /usr/bin/slirp4netns
    Package: slirp4netns-0.4.0-20.1.dev.gitbbd6f25.fc31.x86_64
    Version: |-
      slirp4netns version 0.4.0-beta.3+dev
      commit: bbd6f25c70d5db2a1cd3bfb0416a8db99a75ed7e
  uptime: 16h 57m 48.44s (Approximately 0.67 days)
registries:
  search:
  - docker.io
  - registry.fedoraproject.org
  - registry.access.redhat.com
  - registry.centos.org
  - quay.io
store:
  ConfigFile: /home/thoraxe/.config/containers/storage.conf
  ContainerStore:
    number: 0
  GraphDriverName: overlay
  GraphOptions:
    overlay.mount_program:
      Executable: /usr/bin/fuse-overlayfs
      Package: fuse-overlayfs-0.7.5-2.fc31.x86_64
      Version: |-
        fusermount3 version: 3.6.2
        fuse-overlayfs: version 0.7.5
        FUSE library version 3.6.2
        using FUSE kernel interface version 7.29
  GraphRoot: /home/thoraxe/.local/share/containers/storage
  GraphStatus:
    Backing Filesystem: extfs
    Native Overlay Diff: "false"
    Supports d_type: "true"
    Using metacopy: "false"
  ImageStore:
    number: 10
  RunRoot: /run/user/1000/containers
  VolumePath: /home/thoraxe/.local/share/containers/storage/volumes
```

**Package info (e.g. output of `rpm -q podman` or `apt list podman`):**

```
podman-1.7.0-2.fc31.x86_64
```

**Additional environment details (AWS, VirtualBox, physical, etc.):**
N/A

**Comment by vrothberg** [comment]
@TomSweeneyRedHat PTAL

**Comment by TomSweeneyRedHat** [comment]
I've just verified.  Copying a file using a relative directory like that is a security risk and not allowed.  We did have a better error message there about this at one point, but it looks like it's been obscured.  I'll see if I can fix that up.

FWIW, here's a much better error message from Docker upstream using a similar setup of copying a file from the directory above the context directory:

```
Step 2/2 : COPY ../upperfile.txt /
COPY failed: Forbidden path outside the build context: ../upperfile.txt ()
```

**Comment by TomSweeneyRedHat** [comment]
Just as a follow up, the issue is arising [here](https://github.com/containers/buildah/blob/master/imagebuildah/stage_executor.go#L439).  The call to SecureJoin with a src of `../upperfile.txt` is not creating an error that would then be caught two lines below with an appropriate error message.  Instead, it's removing the `../` and returning `upperfile.txt` which is later appended to the context directory.  When the system then looks for `{context-directory}/upperfile.txt` it fails as the file lives a directory above.

I've asked @cyphar at https://github.com/cyphar/filepath-securejoin/issues/2 what the expectation is for the SecureJoin function.  I think it should be raising an error.  If it does not, I'll adjust the Buildah code to handle the situation better.

**Comment by cyphar** [comment]
I'll post a response both there and here -- the reasoning is that `SecureJoin` (which is soon going to be out-dated by `libpathrs`'s path handling which is race-free) is effectively acting like a `chroot`. Thus `/..` always becomes `/`.

This is necessary because there are lots of examples of malformed symlinks which only work because of this behaviour. If we made `/..` cause an error, those symlinks wouldn't work -- and it's also unclear whether absolute symlinks should also trigger the error (in the `podman build` example, the symlink will be resolved differently to the host -- and this may be confusing to some users -- but on the other hand we absolutely need to allow absolute symlinks to work because many distributions use absolute symlinks all over the place).

**Comment by vrothberg** [comment]
> aside from the fact that SecureJoin is going to be archived soon (once we switch enough people to libpathrs)

@cyphar, do you have a timeline for that? Libpathrs requires new kernels, which forces us to use `SecureJoin` for a longer time, right? It would be great if Buildah could continue using `SecureJoin` but we had to address the stripping of `../`.

**Comment by cyphar** [comment]
@vrothberg libpathrs does **not** require new kernels -- it does (safe) user-space emulation for old kernels, precisely to avoid the concern about bumping the kernel version to 5.6. There are some security hardenings it's not possible to do on old kernels (or even at all right now) but libpathrs will always work best-effort on old kernels.

As for a timeline, I recently merged Go bindings for libpathrs, and am planning for it to be ready for other projects to port to (such as `umoci`) in a few months (I'm going on a vacation soon so that's the main reason for the timeline being longer than a month or two). But once I sort out a few remaining bug, you'll be able to start the porting effort -- the core API won't change much after that.

**Comment by vrothberg** [comment]
> @vrothberg libpathrs does not require new kernels -- it does (safe) user-space emulation for old kernels, precisely to avoid the concern about bumping the kernel version to 5.6

Perfect! Thanks for the explanation.

> I'm not sure what the right solution is for that problem (and libpathrs will have a similar problem if you want to have semantics where ../ is treated specially -- in fact, given that on newer kernels the resolution will be done in-kernel it will be even harder). But I'd be happy to work with you to figure that out.

I begin to think we should address the issue on Buildah's side. We somehow regressed but for good reasons. We could improve the error message (and docs?) to better communicate what's happening. @TomSweeneyRedHat, @rhatdan  WDYT?

**Comment by cyphar** [comment]
@vrothberg 

> I begin to think we should address the issue on Buildah's side. We somehow regressed but for good reasons. We could improve the error message (and docs?) to better communicate what's happening.

I agree (since the theoretical usecase of these safe resolution tools is more for container runtimes operating on a container rootfs than the whole "build context" concept from `docker build`), but feel free to ping me if you want my input.

**Comment by rhatdan** [comment]
sgtm

**Comment by rhatdan** [comment]
@TomSweeneyRedHat Since the buildah got merged, can we close this now?  Has it been vendored into podman?

**Comment by mheon** [comment]
I'm fairly certain we're waiting on your containers.conf for a new Buildah vendor

**Comment by TomSweeneyRedHat** [comment]
As @mheon surmised.  We're on hold vendoring Buildah until the containers.conf is fixed.  This has been fixed in Buildah upstream, but on hold here.

**Comment by mheon** [comment]
We definitely have a new Buildah landed since this was written, so closing
